Issue/2076 user registration add on allow user activation on the view

omarkasem opened the pull request:
- Implements https://github.com/GravityKit/GravityView/issues/2076
- Adds User Activation field type to GravityView that shows ONLY WHEN User Activation is set to Manual
- Makes it work using ajax in multiple and single views

💾 [Build file](https://www.dropbox.com/scl/fi/unlqo74nnxaf3n7tk6zej/gravityview-2.32-a0e94414b.zip?rlkey=56k1uhus33smpxjiigydi5dha&dl=1) (a0e94414b).

mrcasual commented on templates/fields/field-user_activation.php:
```diff
@@ -0,0 +1,42 @@
+?>
+
+<a href="#" activation-key="<?php echo $activation_key; ?>" class="button gv-user-activation-link" style="cursor: pointer;">
+	<?php esc_html_e( 'Activate User', 'gk-gravityview' ); ?>
```
@omarkasem, please use `esc_attr_e`.

mrcasual commented:
@Mwalek, please test.

Mwalek commented:
@mrcasual I found a few issues:

1. There is the following warning in the front end if you create a View and attach a form that is not a user registration form or don't create a user registration feed.`Warning: Undefined array key "is_active" in /.../wp-content/plugins/gravityview/includes/fields/class-gravityview-field-user-activation.php on line 149`
2. There is a fatal error if the View is visited after deactivating the User Registration add on: `Fatal error: Uncaught Error: Call to undefined function gf_user_registration() in /.../wp-content/plugins/gravityview/templates/fields/field-user_activation.php:15`
3. The activate button does not appear in the multiple entries listing if the user is not activated and the user activation field is added to both multiple entries and single entry tab.
<img width="638" alt="Screenshot 2024-10-31 at 6 01 01 PM" src="https://github.com/user-attachments/assets/e3d8487c-349d-43ec-9a5e-7d94cf41cf72">

## Summary
- [ ] Warning on non-registration forms: Undefined array key "is_active".
- [ ] Fatal error if User Registration add-on is deactivated.
- [ ] Activate button missing for unactivated users in multiple/single entry tabs.

mrcasual commented:
@omarkasem, please address the QA issues. Thank you.

omarkasem commented:
@Mwalek 
1,2 Done
3- I can't recreate that, I tried with table, list and it's working for me and showing.

rafaehlers commented:
I'll test it!

omarkasem commented:
@Mwalek Thanks for noticing that, Both are fixed now with the latest commit

Mwalek commented:
@omarkasem works like a charm, thank you!

cc. @mrcasual

rafaehlers commented:
I don't see the point in having "Invalid activation key." for existing users that were previously activated in some way. This appears like an error, and you know, if you display something that looks like an error, the users will ask about it. 

Would it be too costly (performance wise) to check every entry against existing users and in case they exist (as users) just display the same "The user is already active." message? In the case of my screenshot below, these two entries belong are associated with users created by the UR feed in the past. I'm not sure if there's a link we can establish to check that. Check if the user listed as the Entry Creator exists, perhaps? 

Also, what is our text style for these messages? I don't like these messages ending with a period. Is that correct?
Before: `The user is already active.` 
After: `The user is already active`

![image](https://github.com/user-attachments/assets/ac2d7f71-134f-4261-8648-e58fbcfa43e5)

omarkasem commented:
@rafaehlers Good point, I added a check now to see if the user already exist before the check for activation key and i also removed the dots at the end of line.

Please test and let me know what you think

mrcasual commented on includes/fields/class-gravityview-field-user-activation.php:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	/**
+	 * Check if the user exists
```
@omarkasem, please review https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#documenting-tips and update all docblocks.

mrcasual commented on templates/fields/field-user_activation.php:
```diff
@@ -0,0 +1,53 @@
+}
+
+if ( ! class_exists( 'GFUserSignups' ) ) {
+	echo __( 'GFUserSignups class does not exist', 'gk-gravityview' );
```
@omarkasem, this should likely be logged as a warning or error rather than directly displayed to the user. For the user-facing message, consider displaying something more generic and user-friendly.

mrcasual commented on assets/js/field-user-activation.js:
```diff
@@ -0,0 +1,72 @@
+                if ( ! response.success ) {
+
+                    alert( response.data.message );
+                    spinner.destroy();
```
@omarkasem, `spinner.destroy()` can be moved outside of the `if` statement since it always applies.

mrcasual commented on includes/fields/class-gravityview-field-user-activation.php:
```diff
@@ -0,0 +1,190 @@
+ * @subpackage GravityView_Field_User_Activation
+ */
+class GravityView_Field_User_Activation extends GravityView_Field {
+
```
@omarkasem, please remove empty lines after class/method/function declaration.

mrcasual commented on includes/fields/class-gravityview-field-user-activation.php:
```diff
@@ -0,0 +1,190 @@
+				'spinner_url'     => GFCommon::get_base_url() . '/images/spinner.svg',
+			)
+		);
+
```
@omarkasem, no empty line here and in similar places elsewhere...

mrcasual commented on includes/fields/class-gravityview-field-user-activation.php:
```diff
@@ -0,0 +1,190 @@
+	 */
+	public static function check_if_user_exist( $form, $entry ) {
+		foreach ( $form['fields'] as $field ) {
+			if ( $field->type === 'email' ) {
```
@omarkasem, check if it's not an email and `continue`.

omarkasem commented:
@mrcasual Thanks for the notes, they are all solved now
